relay/circuit-v2.md

```diff
@@ -0,0 +1,193 @@
+
+Compared to the first version of the protocol, there are some significant departures:
+- The protocol has been split into two subprotocols, `hop` and `stop`
+  - The `hop` protocol governs the behaviour of relays; it is used for
+    reserving resources in the relay and opening a switched connection
+    to a peer through the relay.
+  - The `stop` protocol governs the termination of circuit switched
+    connections.
+- The concept of resource reservation has been introduced, whereby
+  peers wishing to use a relay explicitly reserve resources and obtain
+  _reservation vouchers_ which can be distributed to their peers for
+  routing purposes.
+- The concept of limited relaying has been introduced, whereby relays
+  provide switched connectivity with a limited duration and data cap.
+
+### Rationale
+
+The evolution of the protocol towards v2 has been influenced by our
+experience in operating open relays in the wild.  The original
+protocol, while very flexible, has some limitations when it comes to
+the practicalities of relaying connections.
+
+The main problem is that is no mechanism to reserve resources in the
+relay, which leads to continuoues oversubscription of relays and the
+necessity of (often inefective) heuristics for balancing resources.
+In practice, running a relay proved to be an expensive proposition
+requiring dedicated hosts with significant hardware and bandwidth
+costs.  In addition, there is ongoing work in Hole Punching
+coordination for direction connection upgrade through relays, which
+doesn't require an unlimited relay connection.
+
+In order to address the situation and seamlessly support pervasive
+hole punching, we have introduced limited relays and slot
+reservations.  This allows relays to effectively manage their
+resources and provide service at a small scale, thus enabling the
+deployment of an army of relays for extreme horizontal scaling without
+excessive bandwidth costs and dedicated hosts.
+
+Furthermore, the original decision to conflate circuit initiation and
+termination in the same protocol has made it very hard to on provide
+relay service on demand, decoupled with whether _client_ functionality
+is supported by the host.
+
+In order to address this problem, we have splt the protocol into the
+`hop` and `stop` subprotocols. This allows us to always enable the
+client-side functionality in a host, while providing the option to
+later mount the relay service in public hosts, _after_ the
+reachability of the host has been determined through AutoNAT.
+
+## The Protocol
+
+### Interaction
+
+The following diagram illustrates the interaction between three peers,
+_A_, _B_, and _R_, in the course of establishing a relayed connection.
+Peer _A_ is a private peer, which is not publicly reachable; it
+utilizes the services of peer _R_ as the relay.  Peer _B_ is another
+peer who wishes to connect to peer _A_ through _R_.
+
+![Circuit v2 Protocol Interaction](circuit-v2.png)
```

Comment:
I would prefer to use a single diagram tool across the repository. We are using plantuml.com/ in https://github.com/libp2p/specs/tree/master/connections, checking in both the svg and the code. This makes it very easy to alter the diagram later on.

@vyzo would you be OK with using plantuml.com/ in this specification?

Reply:
I used excalidraw mostly because I like it, but I am fine with any tool really.

Reply:
Feel free to recreate the diagram with another tool.

Reply:
:+1: https://github.com/libp2p/specs/pull/326 does the change.